[aot] refactor runtime_wrapper's epilogue args access

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122535
* #123359
* #122353
* __->__ #123674
* #123630

I want runtime_wrapper args to be stealable by call_func_at_runtime_with_args, since the args may contain activations which we don't want to hold alive in this scope.

The args to runtime_wrapper **should always be** from a list created within aot_autograd, so it **should always be** safe to steal them: https://github.com/pytorch/pytorch/blob/a4a49f77b8c45ea459263c2242ab391b3d0577f2/torch/_functorch/aot_autograd.py#L928-L932

There are some accesses after we execute the compiled_fn, but those index accesses are already inferred at compile time.

xmfan committed:
[aot] refactor runtime epilogue args access

[ghstack-poisoned]

xmfan committed:
Update on "[aot] refactor runtime epilogue args access"




[ghstack-poisoned]

xmfan committed:
Update on "[aot] refactor runtime_wrapper's epilogue args access"


I want runtime_wrapper args to be stealable by call_func_at_runtime_with_args, since the args may contain activations which we don't want to hold alive in this scope.

The args to runtime_wrapper **should always be** from a list created within aot_autograd, so it **should always be** safe to steal them: https://github.com/pytorch/pytorch/blob/a4a49f77b8c45ea459263c2242ab391b3d0577f2/torch/_functorch/aot_autograd.py#L928-L932

There are some accesses after we execute the compiled_fn, but those index accesses are already inferred at compile time.


[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -72,15 +72,29 @@ def create_runtime_wrapper(
 
+    # args that are still used after we run compiled function (maybe w/ steal_args)
+    epilogue_args_idx = []
+    epilogue_args_idx.extend(runtime_metadata.mutated_inp_runtime_indices)
```
It looks like the idea here is that AOTAutograd's runtime epilogue needs to use the input tensors after the compiled fn finishes, so now that we are doing list clearing here we need to manually keep a reference to those inputs?

It does feel pretty fragile... although at least if we get it wrong, it should error loudly (the runtime epilogue will try to access an index not in `orig_inputs`).

Do you mind adding a longer comment describing the situation here? Something like:
```
Note [Inputs needed in runtime epilogue after list clearing]
Compiled Autograd requires us to pass inputs around in a list so that activations can be freed early.
However, we cannot blindly clear the entire list, because AOTAutograd may need access to some of the graph inputs **after** the compiled function has finished running. There are two main cases:
(1) input mutations. If there are an input mutations that we must run outside of the graph, we need access to the input
(2) output aliasing: outputs that aliases graph inputs generally must be regenerated outside of the `autograd.Function`, and doing so requires us accessing the corresonding input after the compiled artifact has run.
```

xmfan committed:
Update on "[aot] refactor runtime_wrapper's epilogue args access"


I want runtime_wrapper args to be stealable by call_func_at_runtime_with_args, since the args may contain activations which we don't want to hold alive in this scope.

The args to runtime_wrapper **should always be** from a list created within aot_autograd, so it **should always be** safe to steal them: https://github.com/pytorch/pytorch/blob/a4a49f77b8c45ea459263c2242ab391b3d0577f2/torch/_functorch/aot_autograd.py#L928-L932

There are some accesses after we execute the compiled_fn, but those index accesses are already inferred at compile time.


[ghstack-poisoned]

xmfan commented:
@pytorchbot merge
